src/Xrefcheck/System.hs

```diff
@@ -36,21 +43,78 @@ readingSystem = unsafePerformIO
 -- Check the respective env variable which is usually set in all CIs.
 askWithinCI :: IO Bool
 askWithinCI = lookupEnv "CI" <&> \case
-  Just "1"                       -> True
+  Just "1" -> True
   Just (map C.toLower -> "true") -> True
-  _                              -> False
+  _ -> False
+
+-- | A FilePath that has been canonicalized with 'System.Directory.canonicalizePath',
+-- and with indirections expanded as much as possible. Should be created via
+-- 'mkGlobPattern'.
+newtype CanonicalPath = UnsafeCanonicalPath
+  { unCanonicalPath :: FilePath
+  } deriving newtype (Show, Eq, Ord)
+
+canonicalizePath :: FilePath -> IO CanonicalPath
+canonicalizePath = fmap canonicalize . Directory.makeAbsolute
+  where
+  canonicalize :: FilePath -> CanonicalPath
+  canonicalize = UnsafeCanonicalPath
+    . expandIndirections
+    . FP.normalise
+    . FP.dropTrailingPathSeparator
```

Comment:
What do you think about do not use `System.Directory.canonicalizePath` and instead implement a weaker version like this one? In this way, we have more control over its behaviour.

Reply:
Looks reasonable, we want to avoid the casing change that Windows does.

Could you explain how did you get to exactly this implementation? I want to be sure that we do not miss anything important that was done by `System.FilePath.canonicalizePath`.

Reply:
I used some of the transformations that are mentioned in the `System.Directory.canonicalizePath` [documentation](https://hackage.haskell.org/package/directory-1.3.8.0/docs/System-Directory.html#v:canonicalizePath). I hope to not have missed anything important, but at least it is hidden behind this smart constructor and we can modify it later.

Reply:
Okay, I've read the sources and can agree that we don't miss anything, except for symlinks resolution that `canonicalizePath` performed.

Not like we need it, we actually better go without it. But now symlinks are not transparent for us anymore, and we have to thoroughly verify how do we handle them. Please create a ticket for this under 3.0.0 milestone.

---

Hilarious actually, `makeAbsolute` and `normalize` seem completely unnecessary, they rather spoil the work of this function than help. But if we remove them, then we effectively use nothing smart from filepath utilities.

Looking forward to seeing the moment when we strip them completely (in a separate PR).

Reply:
For the record: https://github.com/serokell/xrefcheck/issues/244 has been created.